### PR TITLE
feat(qe): Allow quaint::Value to include metainformation about the corresponding native database type

### PR DESCRIPTION
**Pull request by miguelff**
Close https://github.com/prisma/team-orm/issues/410

This patch overcomes a long-time limitation in quaint, which is that it's main model was an enum, which required chaning all variants to include data common to all types.

We need this because in driver adapters, it's not possible to achieve full correctness unless we implement conversions similarly to those of quiant, for instance, in postgres, we convert data types based on a combination of the prisma data type for a scalar, and its native type.

```rust
         (Value::DateTime(value), &PostgresType::TIME) => value.map(|value| value.time().to_sql(ty, out)),
            (Value::DateTime(value), &PostgresType::TIMETZ) => value.map(|value| {
                let result = value.time().to_sql(ty, out)?;
                // We assume UTC. see https://www.postgresql.org/docs/9.5/datatype-datetime.html
                out.extend_from_slice(&[0; 4]);
                Ok(result)
            }),
```

This is possible in rust quaint because the postgres rust driver has an in memory cache of the database column types. But in driver adapters we don't have that information available and we need to provide it in the quaint representation of a query, and with the current design of a `quaint::Value` that is not possible.

In summary, this PR allows for correct data type conversions for driver adapters which otherwise won't be possible.

### CODE REVIEW

File: quaint/src/visitor/postgres.rs [hunk 1]
```diff
@@ -861,7 +863,7 @@ mod tests {
             vec![serde_json::json!({"a": "b"})],
         );
 
-        let value_expr: Expression = Value::json(serde_json::json!({"a":"b"})).into();
```
Comment (miguelff) [inline quaint/src/visitor/postgres.rs:864]: use factory

File: quaint/src/visitor/postgres.rs [hunk 2]
```diff
@@ -891,7 +893,7 @@ mod tests {
...
+        let value_expr: Expression = ValueType::json(serde_json::json!({"a":"b"})).into();
```
Comment (miguelff) [inline quaint/src/visitor/postgres.rs:896]: use factory

File: quaint/src/visitor/postgres.rs [hunk 3]
```diff
@@ -1007,7 +1009,7 @@ mod tests {
 
     #[test]
     fn test_raw_char() {
-        let (sql, params) = Postgres::build(Select::default().value(Value::character('a').raw())).unwrap();
+        let (sql, params) = Postgres::build(Select::default().value(ValueType::character('a').raw())).unwrap();
```
Comment (miguelff) [inline quaint/src/visitor/postgres.rs:1012]: can this receive value directly?

**Review by miguelff — COMMENTED** [review]
Note to self:

File: quaint/src/ast.rs [hunk 4]
```diff
@@ -54,4 +54,4 @@ pub use table::*;
 pub use union::Union;
 pub use update::*;
 pub(crate) use values::Params;
-pub use values::{IntoRaw, Raw, Value, Values};
+pub use values::{IntoRaw, Raw, Value, ValueType, Values};
```
Comment (miguelff) [inline quaint/src/ast.rs:57]: Unfortunately this is required for pattern matching

File: quaint/src/ast/enums.rs [hunk 5]
```diff
@@ -14,11 +14,14 @@ impl<'a> EnumVariant<'a> {
...
-        Value::Text(Some(self.0))
+        Value::text(self.0)
     }
 
     pub fn into_enum(self, name: Option<EnumName<'a>>) -> Value<'a> {
-        Value::Enum(Some(self), name)
+        match name {
+            Some(name) => Value::enum_variant_with_name(self.0, name),
+            None => Value::enum_variant(self.0),
+        }
```
Comment (miguelff) [inline quaint/src/ast/enums.rs:24]: Opted for being explicit rather than one single method with an optional argument

**Comment by miguelff** [comment]
Something really weird that's happening, is that the [tests that are failing](https://github.com/prisma/prisma-engines/actions/runs/6407174972/job/17393407078) point to some code that doesn't exist anymore. I'm investigating. 

Screenshot 2023-10-04 at 16 19 32

File: .github/workflows/query-engine-driver-adapters.yml [hunk 6]
```diff
@@ -61,14 +61,6 @@ jobs:
...
           echo "STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV
 
-      - uses: actions/cache@v3
-        name: "Setup pnpm cache"
-        with:
-          path: ${{ env.STORE_PATH }}
-          key: ${{ runner.os }}-pnpm-store-${{ hashFiles('**/pnpm-lock.yaml') }}
-          restore-keys: |
-            ${{ runner.os }}-pnpm-store-
-
```
Comment (jkomyno) [inline .github/workflows/query-engine-driver-adapters.yml:71]: Any particular reason or context for this change?

File: quaint/src/connector/metrics.rs [hunk 7]
```diff
@@ -36,7 +36,7 @@ where
 
             trace_query(&query_fmt, params, result, start);
         } else {
-            trace_query(&query, params, result, start);
+            trace_query(query, params, result, start);
```
Comment (jkomyno) [inline quaint/src/connector/metrics.rs:39]: Unintended change?

File: schema-engine/sql-schema-describer/src/postgres.rs [hunk 8]
```diff
@@ -663,13 +663,7 @@ impl<'a> SqlSchemaDescriber<'a> {
...
         "#;
 
-        let rows = self
-            .conn
-            .query_raw(
-                sql,
-                &[Array(Some(namespaces.iter().map(|v| v.as_str().into()).collect()))],
-            )
-            .await?;
+        let rows = self.conn.query_raw(sql, &[Value::array(namespaces)]).await?;
```
Comment (jkomyno) [inline schema-engine/sql-schema-describer/src/postgres.rs:666]: Nice simplification (here and in the following changes to this file).

File: query-engine/connectors/sql-query-connector/src/model_extensions/scalar_field.rs [hunk 9]
```diff
@@ -43,21 +43,21 @@ impl ScalarFieldExt for ScalarField {
...
+            (PrismaValue::Bytes(b), _) => Value::bytes(b),
             (PrismaValue::Object(_), _) => unimplemented!(),
             (PrismaValue::Null, ident) => match ident {
-                TypeIdentifier::String => Value::Text(None),
-                TypeIdentifier::Float => Value::Numeric(None),
-                TypeIdentifier::Decimal => Value::Numeric(None),
-                TypeIdentifier::Boolean => Value::Boolean(None),
+                TypeIdentifier::String => Value::null_text(),
+                TypeIdentifier::Float => Value::null_numeric(),
+                TypeIdentifier::Decimal => Value::null_numeric(),
```
Comment (jkomyno) [inline query-engine/connectors/sql-query-connector/src/model_extensions/scalar_field.rs:59]: Small nit:

```suggestion
                TypeIdentifier::Float | TypeIdentifier::Decimal => Value::null_numeric(),
```

File: query-engine/connector-test-kit-rs/query-tests-setup/src/config.rs [hunk 10]
```diff
@@ -215,7 +215,11 @@ impl TestConfig {
             #[cfg(unix)]
             {
                 use std::os::unix::fs::PermissionsExt;
-                if path.metadata().is_ok_and(|md| md.permissions().mode() & 0o111 == 0) {
+                let is_executable = match path.metadata() {
+                    Err(_) => false,
+                    Ok(md) => md.permissions().mode() & 0o111 != 0,
+                };
+                if !is_executable {
```
Comment (jkomyno) [inline query-engine/connector-test-kit-rs/query-tests-setup/src/config.rs:222]: Not necessarily a relevant change to this PR, but it's ok :)

File: query-engine/driver-adapters/src/proxy.rs [hunk 11]
```diff
@@ -615,167 +617,167 @@ mod proxy_test {
     use super::*;
 
     #[track_caller]
-    fn test_null(quaint_none: QuaintValue, column_type: ColumnType) {
+    fn test_null<'a, T: Into<QuaintValue<'a>>>(quaint_none: T, column_type: ColumnType) {
```
Comment (jkomyno) [inline query-engine/driver-adapters/src/proxy.rs:620]: Nice generic here.

File: query-engine/connector-test-kit-rs/query-tests-setup/src/config.rs (hunk 10 above)
Comment (miguelff) [reply query-engine/connector-test-kit-rs/query-tests-setup/src/config.rs:222]: Yes, I discovered the previous construct was experimental :/ as much as I liked it better. Good catch!

File: .github/workflows/query-engine-driver-adapters.yml (hunk 6 above)
Comment (miguelff) [reply .github/workflows/query-engine-driver-adapters.yml:71]: I was seeing weird errors yesterday that might have been caused by caching. I noticed the runtime of tests didn't grow significantly and the problems ceased (coincidentally) when I removed this block.

File: quaint/src/ast/values.rs [hunk 12]
```diff
@@ -34,11 +35,464 @@ where
...
+        ValueType::int64(value).into_value()
+    }
+
+    /// Creates a new 32-bit signed integer.
+    pub fn integer<I>(value: I) -> Self
+    where
+        I: Into<i32>,
+    {
+        ValueType::int32(value).into_value()
+    }
```
Comment (jkomyno) [inline quaint/src/ast/values.rs:67]: Isn't this redundant? We have `int32` already, with exactly the same implementation.

File: quaint/src/ast/values.rs [hunk 13]
```diff
@@ -34,11 +35,464 @@ where
...
+
+    /// Returns an `i32` if the value is a 32-bit signed integer, otherwise `None`.
+    pub fn as_i32(&self) -> Option<i32> {
+        self.typed.as_i32()
+    }
+
+    /// Returns an `i64` if the value is a signed integer, otherwise `None`.
+    pub fn as_integer(&self) -> Option<i64> {
+        self.typed.as_integer()
+    }
```
Comment (jkomyno) [inline quaint/src/ast/values.rs:270]: Similarly to above, `integer` is redundant compared to `i32`.

File: quaint/src/ast/values.rs (hunk 12 above)
Comment (miguelff) [reply quaint/src/ast/values.rs:67]: Yes, It's redundant, I think we are using aliases from integer to int32. I didn't go that far on the semantics. I can improve that if you think it's a blocker, tho'

File: quaint/src/ast/values.rs [hunk 14]
```diff
@@ -212,304 +672,316 @@ impl<'a> From<Value<'a>> for serde_json::Value {
...
+            Self::Text(Some(cow)) => Some(cow.into_owned()),
+            Self::Bytes(Some(cow)) => String::from_utf8(cow.into_owned()).ok(),
             _ => None,
         }
     }
 
     /// Returns whether this value is the `Bytes` variant.
-    pub const fn is_bytes(&self) -> bool {
-        matches!(self, Value::Bytes(_))
+    pub(crate) fn is_bytes(&self) -> bool {
```
Comment (jkomyno) [inline quaint/src/ast/values.rs:891]: Why do we lose the `const` here (and below)?

File: quaint/src/ast/values.rs (hunk 12 above)
Comment (jkomyno) [reply quaint/src/ast/values.rs:67]: I don't think this is a blocker, but the smaller the API, the better imho. And if we have just identified this redundancy, we might as well say goodbye to that `integer` method (and related `is_integer` and `as_integer` ones) now that it's fresh in our memory :)
Not a blocker at all, though.

**Review by jkomyno — APPROVED** [review]
Thanks for the PR. It's very large, but I do understand it's basically a new `typed` field addition propagated over dozens of different types and pattern matching scenarios.

I've left a few comments that are non binding to the merge.

File: quaint/src/ast/values.rs [hunk 15]
```diff
@@ -34,11 +35,464 @@ where
...
+        ValueType::character(value).into_value()
+    }
+
+    /// Creates a new array value.
+    pub fn array<I, V>(value: I) -> Self
+    where
+        I: IntoIterator<Item = V>,
+        V: Into<Value<'a>>,
+    {
+        ValueType::array(value).into_value()
```
[the remark this answers is not in the document]
Comment (miguelff) [reply quaint/src/ast/values.rs:158]: Correct, this is because the ValueType::Array` variant is a tuple struct of `Option<Vec<Value<'a>>>`

File: quaint/src/ast/values.rs (hunk 14 above)
Comment (miguelff) [reply quaint/src/ast/values.rs:891]: All the methods, before guaranteed stack allocation. Those of the proxy don't anymore. We could make them const, but if later we use them in a constant scope, like a constant initialization, they will have to remain constant and their implementation won't be allowed to vary.  Given proxy methods in value cannot be const, I decided to dowgrade these as well as non-const.

**Comment by miguelff** [comment]
Thank you for the thorough review, @jkomyno . I couldn't split it easily and it ended up being large indeed.

File: quaint/src/ast/values.rs (hunk 12 above)
Comment (miguelff) [reply quaint/src/ast/values.rs:67]: Agreed :) Let me fix it.

File: quaint/src/ast/values.rs (hunk 14 above)
Comment (jkomyno) [reply quaint/src/ast/values.rs:891]: Ok then, thanks for providing the missing context.

File: query-engine/connectors/sql-query-connector/src/row.rs [hunk 16]
```diff
@@ -92,12 +92,13 @@ impl ToSqlRow for ResultRow {
...
+                (type_identifier, FieldArity::List) => match p_value.typed {
                     value if value.is_null() => Ok(PrismaValue::List(Vec::new())),
-                    Value::Array(None) => Ok(PrismaValue::List(Vec::new())),
-                    Value::Array(Some(l)) => l
+                    ValueType::Array(None) => Ok(PrismaValue::List(Vec::new())),
+                    // TODO: avoid cloning in here.
+                    ValueType::Array(Some(l)) => l
                         .into_iter()
-                        .map(|p_value| row_value_to_prisma_value(p_value, meta[i]))
+                        .map(|val| row_value_to_prisma_value(val.clone(), meta[i]))
```
Comment (aqrln) [inline query-engine/connectors/sql-query-connector/src/row.rs:101]: Why do we need to clone here? This compiles as well:
```suggestion
                        .map(|val| row_value_to_prisma_value(val, meta[i]))
```

Comment (miguelff) [reply query-engine/connectors/sql-query-connector/src/row.rs:101]: We don't this was a left over from an intermediate state. I believe. Fixing.

File: quaint/src/ast/values.rs [hunk 17]
```diff
@@ -54,7 +508,7 @@ pub enum Value<'a> {
     /// Read more about it here: https://github.com/prisma/prisma-engines/pull/4280
     Enum(Option<EnumVariant<'a>>, Option<EnumName<'a>>),
     /// Database enum array (PostgreSQL specific).
-    /// We use a different variant than `Value::Array` to uplift the `EnumName`
+    /// We use a different variant than `ValueInner::Array` to uplift the `EnumName`
```
Comment (aqrln) [inline quaint/src/ast/values.rs:511]: should this be `ValueType::Array`?

File: quaint/src/ast/values.rs (hunk 12 above)
Comment (miguelff) [reply quaint/src/ast/values.rs:67]: `is_integer` and `as_integer` cannot be removed. They match on both `ValueType::Int32` and `ValueType::Int64`

Comment (jkomyno) [reply quaint/src/ast/values.rs:67]: Ok, my bad.

File: quaint/src/ast/values.rs (hunk 13 above)
Comment (miguelff) [reply quaint/src/ast/values.rs:262]: No, the implementation in "ValueType" differs from that of i32.

File: quaint/src/ast/values.rs [hunk 18]
```diff
@@ -34,11 +35,464 @@ where
...
+
+    /// `true` if the `Value` is a 64-bit signed integer.
+    pub fn is_i64(&self) -> bool {
+        self.typed.is_i64()
+    }
+
+    /// `true` if the `Value` is a signed integer.
+    pub fn is_integer(&self) -> bool {
+        self.typed.is_integer()
+    }
```
[the remark this answers is not in the document]
Comment (miguelff) [reply quaint/src/ast/values.rs:247]: No, the implementation on `ValueType` differs from i32.
